### PR TITLE
Performance Runbook for Graviton instances

### PR DESCRIPTION
**Pull request by geoffreyblake**
Create a performance runbook that contains steps and best practices for testing, debugging and optimizing application code for AWS Graviton instances.  This PR also contains helper scripts for performing some of the common tasks.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: perfrunbook/debug_hw_perf.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# Debugging performance — “What part of the hardware is slow?”
+
+[Graviton Performance Runbook toplevel](./graviton_perfrunbook.md)
+
+Sometimes it is not code but the hardware that is not performing as well as it should. This may show up in the on-cpu profiles as every function is slightly slower on Graviton as more CPU time is consumed, but no obvious hot-spot function exists.  If this is the case, then measuring how the hardware performs can offer insight.  To do this requires measuring special counters in the CPU to understand which component of the CPU is bottlenecking the code from executing as fast as possible.
```
Comment (AGSaidi) [inline perfrunbook/debug_hw_perf.md:5]: should is a strong word

File: perfrunbook/utilities/capture_flamegraphs.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+help_msg() {
+ echo "usage: $0 oncpu|offcpu [time seconds|default 300]"
+}
+
+capture_on_cpu () {
+  perf record -a -g -k 1 -F99 -e cpu-clock:pppH -- sleep $1
+  perf inject -j -i perf.data -o perf.data.jit
+  perf script -f -i perf.data.jit > script.out
+  ./FlameGraph/stackcollapse-perf.pl --kernel --jit script.out > folded.out
```
Comment (AGSaidi) [inline perfrunbook/utilities/capture_flamegraphs.sh:11]: where is this expected to be run from.. perhaps you want to use `$(dirname $0)`here to find the directory the script is and construct a relative path?

File: perfrunbook/utilities/capture_flamegraphs.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+
+help_msg() {
+ echo "usage: $0 oncpu|offcpu [time seconds|default 300]"
+}
+
+capture_on_cpu () {
+  perf record -a -g -k 1 -F99 -e cpu-clock:pppH -- sleep $1
```
Comment (AGSaidi) [inline perfrunbook/utilities/capture_flamegraphs.sh:8]: you should check that perf exists in the path and is runnable and print help if not

File: perfrunbook/utilities/find_and_list_jar_with_so.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
+if [[ $(id -u) -ne 0 ]]; then
```
Comment (AGSaidi) [inline perfrunbook/utilities/find_and_list_jar_with_so.sh:3]: really, why?

File: perfrunbook/utilities/find_and_list_jar_with_so.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  echo "Must run with sudo privileges"
+  exit 1
+fi
+
+i=$(find / -name '*.jar' -type f -print)
+i=($i)
+
+mkdir -p jar_dir
```
Comment (AGSaidi) [inline perfrunbook/utilities/find_and_list_jar_with_so.sh:11]: not a dir in /tmp?

File: perfrunbook/utilities/find_and_list_jar_with_so.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+cd jar_dir
+rm -rf * jar_dir
```
Comment (AGSaidi) [inline perfrunbook/utilities/find_and_list_jar_with_so.sh:13]: ?

File: perfrunbook/utilities/find_and_list_jar_with_so.sh [hunk 7]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash
+
```
Comment (AGSaidi) [inline perfrunbook/utilities/find_and_list_jar_with_so.sh:2]: `set -e` given the number of `rm *` we don't want something failing and worst happening

File: perfrunbook/utilities/configure_vcpus.sh [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+
```
Comment (AGSaidi) [inline perfrunbook/utilities/configure_vcpus.sh:2]: set -e?

File: perfrunbook/utilities/find_and_list_pylib_with_so.sh [hunk 9]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
```
Comment (AGSaidi) [inline perfrunbook/utilities/find_and_list_pylib_with_so.sh:2]: same as above

File: perfrunbook/utilities/install_perfrunbook_dependencies_al2.sh [hunk 10]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
```
Comment (AGSaidi) [inline perfrunbook/utilities/install_perfrunbook_dependencies_al2.sh:2]: unify into one script that detects redhat-like or debian-like and does the right thing

File: perfrunbook/configuring_your_sut.md [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+8. We recommend using instances set to [**dedicated tenancy**](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/dedicated-instance.html) to control for noisy neighbor affects that can occur when using instances that are set to **default** tenancy and lead to noisy results.
+
+## Check for missing binary dependencies
+
+Libraries for Python or Java can link in binary shared objects to provide enhanced performance.  Lacking these shared object dependencies many times do not prevent the service from running on Graviton2, but will be forced to use a slow code-path instead of the optimized paths.  Use the checklist below to verify the same shared objects are available on all platforms.
+
+1. JVM based languages — Check for the presence of binary shared objects in the installed JARs and compare between Graviton2 and x86.
+  ```bash
+  %> cd ~/aws-getting-started-guide/perfrunbook/utilities
+  %> sudo ./find_and_list_jar_with_so.sh
```
[the remark this answers is not in the document]
Comment (geoffreyblake) [reply perfrunbook/configuring_your_sut.md:84]: It searches the entire system to look where all the JARs may get deployed.

File: perfrunbook/graviton_perfrunbook.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    Begin by verifying software dependencies and verifying the configuration of your Graviton and x86 testing environments to check that no major differences are present in the testing environment.  Performance differences may be due to differences in environment and not the due to the hardware.  Refer to the below chart for a step-by-step flow through this runbook to help root cause the performance regression:
+    ![](./images/performance_debug_flowchart.png)
+* **What are the recommended optimizations to try with Graviton2?**
+    Refer to [Section 6](./optimization_recommendation.md) for our recommendations on how to make your application run faster on Graviton.
+* **I investigated every optimization in this guide and still cannot find the root-cause, what do I do next?**
+    Start by documenting the basics of your application's dependencies, runtime environment, benchmark and optimization steps tried.  Archive your gathered profiles that you obtained following this runbook into a zip file.  Post this as an issue to our [AWS-Graviton-Getting-Started](https://github.com/aws/aws-graviton-getting-started/issues) guide to get assistance with further debugging.
```
[the remark this answers is not in the document]
Comment (geoffreyblake) [reply perfrunbook/graviton_perfrunbook.md:54]: Good point.

File: perfrunbook/utilities/find_and_list_jar_with_so.sh (hunk 5 above)
Comment (geoffreyblake) [reply perfrunbook/utilities/find_and_list_jar_with_so.sh:11]: Its expected to run from the utilities directory, but can move it to /tmp to be cleaner.

File: perfrunbook/utilities/find_and_list_jar_with_so.sh (hunk 6 above)
Comment (geoffreyblake) [reply perfrunbook/utilities/find_and_list_jar_with_so.sh:13]: That's a bug. Fixed in the coming revision.

File: perfrunbook/utilities/find_and_list_jar_with_so.sh (hunk 4 above)
Comment (geoffreyblake) [reply perfrunbook/utilities/find_and_list_jar_with_so.sh:3]: The find command searches the entire disk.  Without sudo, it errors out. I don't want to assume the machine this gets run on deploys artifacts into home, or only to some canned directory.  There are a multitude of places JARs may get placed.

File: README.md [hunk 13]
```diff
@@ -22,6 +22,7 @@ This repository is meant to help new users start using the Arm-based AWS Gravito
 * [TensorFlow](tensorflow.md)
 * [Known issues and workarounds](#known-issues-and-workarounds)
 * [AWS Managed Services available on Graviton](managed_services.md)
+* [Graviton Performance Runbook](perfrunbook/graviton_perfrunbook.md)
```
Comment (sebpop) [inline README.md:25]: Maybe also add a link to the runbook from Section "# Optimizing for Graviton" below.

File: perfrunbook/utilities/find_and_list_jar_with_so.sh (hunk 4 above)
Comment (AGSaidi) [reply perfrunbook/utilities/find_and_list_jar_with_so.sh:3]: i'd expect most to be readable

File: perfrunbook/utilities/configure_graviton_metal_iommu.sh [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+if [[ $(id -u) -ne 0 ]]; then
+  echo "Must be run with sudo privileges"
+  exit 1
+fi
+
+
+
+cp $grub_loc ${grub_loc}.bak
+
+if [[ "$1" == "on" ]]; then
```
[the remark this answers is not in the document]
Comment (geoffreyblake) [reply perfrunbook/utilities/configure_graviton_metal_iommu.sh:44]: I see what you mean.  I'll use the case statement so its cleaner and test it.

**Pull request merged**